### PR TITLE
Simplified gmm clean

### PR DESCRIPTION
**Pull request by charlotte-avery**
New, simplified version of GMM with method taken from sk-learn GMM implementation and made compatible for multiple GPU training using PyTorch Lightning.

### CODE REVIEW

File: src/opensynth/models/faraday/gaussian_mixture/__init__.py [hunk 1]
```diff
@@ -1,9 +1,2 @@
 # Copyright Contributors to the Opensynth-energy Project.
 # SPDX-License-Identifier: Apache-2.0
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/__init__.py:2]: I think we could add some of the modules here to make it more convenient for end users to import I think!

File: src/opensynth/models/faraday/gaussian_mixture/gmm_init.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def initialise_gmm_params(
+    data: torch.Tensor,
+    n_components: int,
+    vae_module: FaradayVAE,
+    reg_covar: float = 1e-6,
+) -> gmm_model.GMMInitParams:
+    """
+    Initialise Gaussian Mixture Parameters. This works
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_init.py:17]: Doc strings are outdated!

File: src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+    def update(self, covariances: torch.Tensor) -> None:
+        self.covariances.add_(covariances)
+
+    def compute(self) -> None:
+        return self.covariances
+
+
+class NLLMetric(Metric):
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py:83]: I think maybe we should called this `LogProbMetric` instead, as this is strictly speaking the log prob, rather than NLL. Unless you want to move the NLL calculation here into this metric, then we call it `NLLMetric`. 

Also I prefer we call it `NegativeLogLikelihoodMetric` too to and avoid abbreviations to make the code a lot more self-descriptive.

File: src/opensynth/models/faraday/gaussian_mixture/gmm_model.py [hunk 4]
```diff
@@ -1,194 +1,421 @@
...
+        self.covariance_metric.reset()
+        self.nll.reset()
 
-            sample = mean + scale
+    def training_step(self, batch) -> None:
+        # Encode the batch
+        encoded_batch = gmm_utils.encode_data(batch, self.vae_module)
 
-            result.append(sample)
+        if "weights" in list(batch.keys()):
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_model.py:322]: I prefer we let end users define the name of the `sample_weight` column. This column might be named differently in different datasets and best we don't presume what the name of the column is.

File: src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py (hunk 3 above)
Comment (shengy90) [reply src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py:83]: Just read further down the line that the value we're passing into this metric is already the negative log likelihood. Would change the attribute `self.log_prob` to `self.nll` instead to make this clearer as the class name doesn't match its attributes.

**Review by shengy90 — CHANGES_REQUESTED** [review]
Left some comments on:

1. Outdated doc strings
2. Update class/ variable name to make things clearer/ more consistent
3. I think you'll also need to expand samples prior to running initialisation
4. Also think about how to pass in sample weights in a more flexible way

File: notebooks/faraday/gmm_batch_implementation_simplified.ipynb [hunk 5]
```diff
@@ -0,0 +1,939 @@
+{
```
Comment (shengy90) [inline notebooks/faraday/gmm_batch_implementation_simplified.ipynb:1]: Do we need this notebook?

File: notebooks/faraday/gmm_implementation_simplified.ipynb [hunk 6]
```diff
@@ -0,0 +1,2047 @@
+{
```
Comment (shengy90) [inline notebooks/faraday/gmm_implementation_simplified.ipynb:1]: Do we need this notebook?

File: src/opensynth/models/faraday/gaussian_mixture/__init__.py (hunk 1 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/__init__.py:2]: Done.

File: src/opensynth/models/faraday/gaussian_mixture/gmm_init.py (hunk 2 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/gmm_init.py]: Updated.

File: src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py (hunk 3 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/gmm_metrics.py]: Updated a suggested.

File: src/opensynth/models/faraday/model.py [hunk 7]
```diff
@@ -235,21 +235,40 @@ def train_gmm(self, dm: LCLDataModule):
                 """
             )
 
-        gmm_module, _, _ = fit_gmm(
-            data=dl,
+        # Fit GMM
+        gmm_init_params = initialise_gmm_params(
```
[the remark this answers is not in the document]
Comment (charlotte-avery) [reply src/opensynth/models/faraday/model.py:239]: Done.

File: notebooks/faraday/gmm_implementation_simplified.ipynb (hunk 6 above)
Comment (charlotte-avery) [reply notebooks/faraday/gmm_implementation_simplified.ipynb:1]: Deleted

File: notebooks/faraday/gmm_batch_implementation_simplified.ipynb (hunk 5 above)
Comment (charlotte-avery) [reply notebooks/faraday/gmm_batch_implementation_simplified.ipynb:1]: Deleted

File: src/opensynth/models/faraday/gaussian_mixture/gmm_init.py [hunk 8]
```diff
@@ -14,23 +14,23 @@ def initialise_gmm_params(
...
-    using K-means, and porting SKLearn's implementation
-    of computing cholesky precision and covariances.
+    Initialise Gaussian Mixture Parameters.
+    Component means and component assignments are initialised using K-means,
+      and porting SKLearn's implementation of computing cholesky precision and
+      covariances.
 
     Args:
-        X (np.array): Input data
+        X (torch.Tensor): Input data
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_init.py:23]: Eh - the Args are not correct! They all form part of the doc strings!

File: src/opensynth/models/faraday/model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:239]: Where is this done? I don't see it in the code..

File: src/opensynth/models/faraday/gaussian_mixture/gmm_init.py [hunk 9]
```diff
@@ -14,23 +14,23 @@ def initialise_gmm_params(
...
         reg_covar (float): Regularisation for covariance matrix
 
     Returns:
         dict[str, torch.Tensor]: GMM params
     """
 
-    # Use data from first batch to initialise centroids
-    # If data is shuffled, this should represent the full dataset
     X = gmm_utils.encode_data(data, vae_module)
+    if "weights" in list(data.keys()):
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_init.py:32]: ♻️ Same comment about `weights`:

- either ask users explicitly for the name of the column, or enumerate it in `TrainingData` as a NotRequired key.

File: src/opensynth/models/faraday/gaussian_mixture/gmm_utils.py [hunk 10]
```diff
@@ -48,6 +48,28 @@ def expand_weights(data: torch.Tensor, weights: torch.Tensor) -> torch.Tensor:
...
+    vae_module: FaradayVAE,
+    sample_weights_column: Optional[str] = None,
+) -> torch.Tensor:
+
+    encoded_data = _encode_data(data, vae_module)
+
+    if sample_weights_column is not None:
+        try:
+            model_input = _expand_weights(
+                encoded_data, data[sample_weights_column]
```
Comment (shengy90) [inline src/opensynth/models/faraday/gaussian_mixture/gmm_utils.py:62]: is `data` an instance of torch.Tensor? Or is this an instance of `TrainingData`?

Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/gmm_utils.py:62]: Actually, yes this should be type `dict[str, torch.Tensor]` (which is the same as TrainingData)

File: tests/models/faraday/gaussian_mixture/test_gmm_inputs.py [hunk 11]
```diff
@@ -22,17 +23,18 @@ class TestGMMDataPreparation:
 
     unweighted_batch = TrainingData(kwh=kwh, features=features)
```
Comment (shengy90) [inline tests/models/faraday/gaussian_mixture/test_gmm_inputs.py:24]: In the line above:
```Python
batch = TrainingData(kwh=kwh, features=features, weights=weights)
```

This shouldn't be allowed. Not sure why mypy has not detected this (see https://peps.python.org/pep-0589/).

The proper way to do this is:

```python
class TrainingDataWithWeights(TrainingData):
    weights: torch.Tensor
```

See this commit:
https://github.com/OpenSynth-energy/OpenSynth/blob/42624dcbb5a612c6058fcfdc55d08c4c8fdd5617/tests/models/faraday/gaussian_mixture/test_gmm_inputs.py

File: src/opensynth/models/faraday/gaussian_mixture/gmm_init.py (hunk 8 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/gaussian_mixture/gmm_init.py]: Updated

File: tests/models/faraday/gaussian_mixture/test_gmm_inputs.py (hunk 11 above)
Comment (charlotte-avery) [reply tests/models/faraday/gaussian_mixture/test_gmm_inputs.py:30]: Updated.
